### DISCUSSION TITLE
Create new action creators?

### DISCUSSION THREAD

**Discussion by geekox86** [statelyai/xstate#3495] [Q&A] [answered]
Hello. I am trying to use `choose` to execute different context assignments in different events. I noticed that even using typegen does not infer the correct event in actions from the filtered event in the condition.

Pure action causes problems with typegen.

Therefore, I am thinking of creating a custom action creator. Is this possible? What do you suggest for my use case?

Thank you.

**Comment by christoph-fricke** [comment, marked as answer]
Hi,

creating a custom action creator is absolutely possible. The key to success is to return an `ActionObject`. For example:

```typescript
export function actionCreator<C, E extends EventObject>(custom: (c: C, e: E) => string, args: string): ActionObject<C, E> {
  // Wrapped in a pure action to avoid runtime warnings about "Unknown action type 'action-name'"
  return pure(() => ({
    type: "action-name",
    exec: (context, event) => {
      // Custom action logic, which might forward the context and event to it's parameters depending on your action.
      const data = custom(context, event);
      console.log(data, args);
    }
  }));
}
```

Am I understanding your use-case correctly: you are using one action for multiple events to assign different context values depending on the event that triggered the action?
Sounds like you would be better off by defining a different action for each assignment action that would be chosen by `choose`.

**Reply by geekox86** [reply]
Hi and thx for passing by. Yes you got my use case right. I tried choose action but it has the problem of not associating the filtered event in the cond expression with the event in the actions to narrow it down and provide exact typing for its data members. That what drove me to think about custom action creator. Although what you have provided as code example is great, I am not sure how I should associate and narrow down the selected event?

**Reply by christoph-fricke** [reply]
Inside `exec` it is all about Typescript. You can use [Narrowing](https://www.typescriptlang.org/docs/handbook/2/narrowing.html), e.g. `if (event.type === "event-name") { ... }`, to select different events.

However, I do not recommend to use a custom action for your use-case and it would not even work. Given that you want to assign values to context, you will have to use the `assign` action, which can not be used successfully inside `exec`, since its returned action object cannot be picked up by the interpreter.

Instead, you have two options (I advocate to use the first option):
1. Creating multiple actions in the machine options, one for each assignment in your previous `choose` approach, and associate the actions with their events in the machine definition. With this approach, Typegen will be able to narrow down the event perfectly without further narrowing required by you.
2. Use `pure` instead of `choose` combined with TypeScript [Narrowing](https://www.typescriptlang.org/docs/handbook/2/narrowing.html). However, after narrowing the event in `pure`, it would still be the union of all possible events in your context assignment functions again, since TS cannot infer that both `pure` and `assign` will receive the same event. You would have to do something like [this](https://github.com/statelyai/xstate/discussions/3494#discussioncomment-3328592). So you might as well use `choose` and apply the approach outlined in the linked discussion.

**Comment by Andarist** [comment]
> Hello. I am trying to use choose to execute different context assignments in different events. I noticed that even using typegen does not infer the correct event in actions from the filtered event in the condition.

> I tried choose action but it has the problem of not associating the filtered event in the cond expression with the event in the actions to narrow it down and provide exact typing for its data members. 

This is an interesting idea - it never has occurred to me. It should be possible to implement something like that in the future but it's certainly not trivial. OTOH, if you need to write a condition anyway I would say that moving the condition to the action's body and relying on the default TS behavior for narrowing unions etc should be an acceptable solution. It's simple, effective and doesn't reinvent the wheel.

> > Am I understanding your use-case correctly: you are using one action for multiple events to assign different context values depending on the event that triggered the action?

> Hi and thx for passing by. Yes you got my use case right

With this Have you considered using transition actions here instead of entry/exit actions? This way your actions within a transition definition should already be narrowed down~ to a particular event type.

**Reply by geekox86** [reply]
Sorry for the delay. Yes I went with nested conditional expressions in the action body.
